[Feature request]  make memfd_create for wayland a optional configure flag

# First and foremost consider this:
- Only RetroArch bugs should be filed here. Not core bugs or game bugs
- This is not a forum or a help section, this is strictly developer oriented

## Description

The Mali T628 Wayland driver for ODroid XU3/XU4 doesn't work with memfd_create.
It will error out with: ```[Wayland] [SHM]: Creating a buffer file for 0 B failed```
Probably because the driver is very dated.
I've removed it from the code:
```
diff --git a/input/common/wayland_common.c b/input/common/wayland_common.c
index 98a01b11c4..25ac084011 100644
--- a/input/common/wayland_common.c
+++ b/input/common/wayland_common.c
@@ -708,33 +708,6 @@ void flush_wayland_fd(void *data)
    }
 }
 
-#ifdef HAVE_MEMFD_CREATE
-int create_anonymous_file(off_t size)
-{
-   int fd;
-
-   int ret;
-
-   fd = memfd_create(SPLASH_SHM_NAME, MFD_CLOEXEC | MFD_ALLOW_SEALING);
-
-   if (fd < 0)
-      return -1;
-
-   fcntl(fd, F_ADD_SEALS, F_SEAL_SHRINK);
-
-   do {
-      ret = posix_fallocate(fd, 0, size);
-   } while (ret == EINTR);
-   if (ret != 0) {
-      close(fd);
-      errno = ret;
-      return -1;
-   }
-
-   return fd;
-}
-#endif
-
 shm_buffer_t *create_shm_buffer(gfx_ctx_wayland_data_t *wl, int width,
    int height,
    uint32_t format)
@@ -747,12 +720,9 @@ shm_buffer_t *create_shm_buffer(gfx_ctx_wayland_data_t *wl, int width,
    stride = width * 4;
    size = stride * height;
 
-#ifdef HAVE_MEMFD_CREATE
-   fd = create_anonymous_file(size);
-#else
    fd = shm_open(SPLASH_SHM_NAME, O_RDWR | O_CREAT, 0660);
    ftruncate(fd, size);
-#endif
+
    if (fd < 0) {
       RARCH_ERR("[Wayland] [SHM]: Creating a buffer file for %d B failed: %s\n",
          size, strerror(errno));
@@ -774,9 +744,8 @@ shm_buffer_t *create_shm_buffer(gfx_ctx_wayland_data_t *wl, int width,
       stride, format);
    wl_buffer_add_listener(buffer->wl_buffer, &shm_buffer_listener, buffer);
    wl_shm_pool_destroy(pool);
-#ifndef HAVE_MEMFD_CREATE
+
    shm_unlink(SPLASH_SHM_NAME);
-#endif
    close(fd);
 
    buffer->data = data;
```
Now it works fine.
### Expected behavior

A nice --enable/disable-wayland-memfd flag :)

### Actual behavior

```[Wayland] [SHM]: Creating a buffer file for 0 B failed```

### Steps to reproduce the bug

1. Just launch retroarch

### Bisect Results

[Try to bisect and tell us when this started happening]

### Version/Commit
You can find this information under Information/System Information

- RetroArch: b697f556e5bf4c02331b5d2701b307b1431f6a99

### Environment information

- OS: Fedora 36 ARM v7
- Compiler: GCC 12.0.1

I also experience this problem in Haswell intel vulkan driver:
`╰─$ retroarch -v                       
[INFO] [Config]: Loading default config.
[INFO] [Config]: Looking for config in: "/home/mario/.config/retroarch/retroarch.cfg".
[INFO] RetroArch 1.10.1 (Git 97c5c269a7)
[INFO] === Build =======================================
[INFO] CPU Model Name: Intel(R) Core(TM) i7-4500U CPU @ 1.80GHz
[INFO] Funcionalidades:  MMX MMXEXT SSE SSE2 SSE3 SSSE3 SSE4 SSE4.2 AES AVX AVX2
[INFO] Built: Mar  5 2022
[INFO] Version: 1.10.1
[INFO] Git: 97c5c269a7
[INFO] =================================================
[INFO] [Input]: Found input driver: "udev".
[INFO] [Environ]: SET_PIXEL_FORMAT: RGB565.
[INFO] [Core]: Versión de la API libretro: 1, API usada en la compilación: 1
[INFO] [Audio]: Set audio input rate to: 48000.00 Hz.
[INFO] [Video]: Set video size to: fullscreen.
[INFO] [Wayland/Vulkan]: Compositor doesn't support zwp_idle_inhibit_manager_v1 protocol
[INFO] [Wayland/Vulkan]: Compositor doesn't support zxdg_decoration_manager_v1 protocol
[ERROR] [Wayland] [SHM]: Creating a buffer file for 0 B failed: Invalid argument
[1]    12855 segmentation fault (core dumped)  retroarch -v
`

--- a/input/common/wayland_common.c
+++ b/input/common/wayland_common.c
@@ -984,32 +984,36 @@ void flush_wayland_fd(void *data)
    }
 }
 
-#ifdef HAVE_MEMFD_CREATE
-int create_anonymous_file(off_t size)
+int create_shm_file(off_t size)
 {
    int fd;
 
    int ret;
 
+   #ifdef HAVE_MEMFD_CREATE
    fd = memfd_create(SPLASH_SHM_NAME, MFD_CLOEXEC | MFD_ALLOW_SEALING);
 
+   if (fd >= 0) {
+     fcntl(fd, F_ADD_SEALS, F_SEAL_SHRINK);
+
+     do {
+        ret = posix_fallocate(fd, 0, size);
+     } while (ret == EINTR);
+     if (ret != 0) {
+        close(fd);
+        errno = ret;
+        fd = -1;
+     }
+   }
    if (fd < 0)
-      return -1;
-
-   fcntl(fd, F_ADD_SEALS, F_SEAL_SHRINK);
-
-   do {
-      ret = posix_fallocate(fd, 0, size);
-   } while (ret == EINTR);
-   if (ret != 0) {
-      close(fd);
-      errno = ret;
-      return -1;
+   #endif
+   {
+      fd = shm_open(SPLASH_SHM_NAME, O_RDWR | O_CREAT, 0660);
+      ftruncate(fd, size);
    }
 
    return fd;
 }
-#endif
 
 shm_buffer_t *create_shm_buffer(gfx_ctx_wayland_data_t *wl, int width,
    int height,
@@ -1023,12 +1027,7 @@ shm_buffer_t *create_shm_buffer(gfx_ctx_wayland_data_t *wl, int width,
    stride = width * 4;
    size = stride * height;
 
-#ifdef HAVE_MEMFD_CREATE
-   fd = create_anonymous_file(size);
-#else
-   fd = shm_open(SPLASH_SHM_NAME, O_RDWR | O_CREAT, 0660);
-   ftruncate(fd, size);
-#endif
+   fd = create_shm_file(size);
    if (fd < 0) {
       RARCH_ERR("[Wayland] [SHM]: Creating a buffer file for %d B failed: %s\n",
          size, strerror(errno));
@@ -1050,9 +1049,7 @@ shm_buffer_t *create_shm_buffer(gfx_ctx_wayland_data_t *wl, int width,
       stride, format);
    wl_buffer_add_listener(buffer->wl_buffer, &shm_buffer_listener, buffer);
    wl_shm_pool_destroy(pool);
-#ifndef HAVE_MEMFD_CREATE
    shm_unlink(SPLASH_SHM_NAME);
-#endif
    close(fd);
 
    buffer->data = data;

--- a/input/common/wayland_common.h
+++ b/input/common/wayland_common.h
@@ -206,7 +206,7 @@ void handle_toplevel_close(void *data,
 
 void flush_wayland_fd(void *data);
 
-int create_anonymous_file(off_t size);
+int create_shm_file(off_t size);
 
 shm_buffer_t *create_shm_buffer(gfx_ctx_wayland_data_t *wl,
    int width, int height, uint32_t format);

